cfbs/index.py

```diff
@@ -0,0 +1,161 @@
+    target = os.path.basename(module)
+    return {
+        "description": "Local policy file added using cfbs command line",
+        "tags": ["local"],
+        "dependencies": ["autorun"],
+        "steps": ["copy %s services/autorun/%s" % (module, target)],
+        "added_by": "cfbs add",
+    }
+
+
+def _local_module_data_json_file(module):
+    return {
+        "description": "Local augments file added using cfbs command line",
+        "tags": ["local"],
+        "steps": ["json %s def.json" % module],
+        "added_by": "cfbs add",
+    }
+
+
+def _local_module_data_subdir(module):
+    assert module.startswith("./")
+    dst = os.path.join("services", "cfbs", module[2:])
+    return {
+        "description": "Local subdirectory added using cfbs command line",
+        "tags": ["local"],
+        "steps": ["directory {} {}".format(module, dst)],
+        "added_by": "cfbs add",
+    }
+
+
+def _generate_local_module_object(module):
+    assert module.startswith("./")
+    assert module.endswith((".cf", ".json", "/"))
+    assert os.path.isfile(module) or os.path.isdir(module)
+
+    if os.path.isdir(module):
+        return _local_module_data_subdir(module)
+    if module.endswith(".cf"):
+        return _local_module_data_cf_file(module)
+    if module.endswith(".json"):
+        return _local_module_data_json_file(module)
+
+
+class Index:
+    """Class representing the cfbs.json containing the index of available modules"""
+
+    def __init__(self, index=_DEFAULT_INDEX):
+        self._unexpanded = index
+        self._data = None
+
+    def __contains__(self, key):
+        return key in self.data["index"]
+
+    def __getitem__(self, key):
+        return self.data["index"][key]
+
+    def items(self):
+        return self.data["index"].items()
+
+    def get(self, key, default=None):
+        return self.data["index"].get(key, default)
+
+    def _expand_index(self):
+        index = self._unexpanded
+        if type(index) in (dict, OrderedDict):
+            self._data = {"type": "index", "index": index}
+            return
+
+        assert type(index) is str
+
+        self._data = get_or_read_json(index)
+
+        if not self._data:
+            sys.exit("Could not download or find module index")
+        if "index" not in self._data:
+            sys.exit("Empty or invalid module index")
+
+    @property
+    def data(self) -> dict:
+        if not self._data:
+            self._expand_index()
+        return self._data
+
+    def exists(self, module):
+        if isinstance(module, dict):
+            name = module["name"]
+            version = module.get("version")
+        else:
+            name = module
+            version = None
+        if os.path.exists(name):
+            return True
+        else:
+            if not version:
+                return name in self
+            else:
+                versions = get_json(_VERSION_INDEX)
+                return name in versions and version in versions[name]
```

Comment:
Early return makes these else's unnecessary:

```suggestion
        if os.path.exists(name):
            return True
        if not version:
            return name in self
        versions = get_json(_VERSION_INDEX)
        return name in versions and version in versions[name]
```

https://refactoring.guru/replace-nested-conditional-with-guard-clauses